### PR TITLE
Adding ISSUE_TEMPLATE

### PR DESCRIPTION
**Pull request by vsiddharth**
This patch adds a template for bug-reports and feature requests.

Signed-off-by: Vinothkumar Siddharth <sidvin@amazon.com>

### Summary

This patch adds a template for bug-reports and feature requests.

### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes:  No

### Description for the changelog

* Template for Issues

### Licensing

This contribution is under the terms of the Apache 2.0 License: 
Yes

### CODE REVIEW

File: .github/ISSUE_TEMPLATE.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+
+### Supporting Log Snippets
+<!-- Please provide supporting log information for this issue -->
+
+
+### Stacktrace
+
+
+### Environment Details
```
Comment (jhaynes) [inline .github/ISSUE_TEMPLATE.md]: We should either give examples or make some demands here. e.g. 

Please provide the out put of:
   `docker info`, 
  `curl http://localhost:51678/v1/metadata | grep Version` 
  etc. 

I think I would move this above the log snippets too.

File: .github/ISSUE_TEMPLATE.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+Uncomment only one of the below items
+-->
+
+<!-- ### Bug Report -->
+<!-- ### Feature Request -->
+
+
+### Summary
+<!-- Please provide a brief outline of the bug or feature request -->
```
Comment (jhaynes) [inline .github/ISSUE_TEMPLATE.md]: Can just call this an issue (since we do below)

File: .github/ISSUE_TEMPLATE.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+### Environment Details
+
+
+### Additional Information
+<!--
+Please have a look at https://github.com/awslabs/ecs-logs-collector
+for the data we typically require for investigation
+-->
+
+<!-- -->
```
Comment (jhaynes) [inline .github/ISSUE_TEMPLATE.md]: Why is this here?

File: .github/ISSUE_TEMPLATE.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
+<!--
+This form is for bug reports and feature requests ONLY!
```
Comment (samuelkarp) [inline .github/ISSUE_TEMPLATE.md]: I think I'd get rid of this line; it's a bit unfriendly and carries the same meaning as the next line about contacting AWS Support.

File: .github/ISSUE_TEMPLATE.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+For high severity issues, please contact AWS Support directly.
+
+Please provide the following information:
+-->
+
+<!--
+BUG REPORT or FEATURE REQUEST ?
+
+Uncomment only one of the below items
```
Comment (samuelkarp) [inline .github/ISSUE_TEMPLATE.md]: I don't think we need this; we can do this triage ourselves when we see the issue.

File: .github/ISSUE_TEMPLATE.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+### Description
+<!-- Provide detailed information about the issue -->
+
+
+<!-- -->
+### Expected Behavior
+
+
+### Supporting Log Snippets
+<!-- Please provide supporting log information for this issue -->
```
Comment (samuelkarp) [inline .github/ISSUE_TEMPLATE.md]: It'd be good to provide instructions for how to find relevant logs.  We can point people to the [ecs-logs-collector](https://github.com/awslabs/ecs-logs-collector) too.

File: .github/ISSUE_TEMPLATE.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+
+### Stacktrace
```
Comment (samuelkarp) [inline .github/ISSUE_TEMPLATE.md]: Where would anyone find a stacktrace?  I think this should be covered by the logs above.

File: .github/ISSUE_TEMPLATE.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+
+### Environment Details
+
+
+### Additional Information
+<!--
+Please have a look at https://github.com/awslabs/ecs-logs-collector
```
Comment (samuelkarp) [inline .github/ISSUE_TEMPLATE.md]: Isn't this the same as the log section above?  Also, it'd be good to actually enumerate the details we want.

File: .github/ISSUE_TEMPLATE.md (hunk 6 above)
Comment (aaithal) [inline .github/ISSUE_TEMPLATE.md]: Please add a pointer to the logs collector.

File: .github/ISSUE_TEMPLATE.md (hunk 8 above)
Comment (vsiddharth) [inline .github/ISSUE_TEMPLATE.md]: No, this isn't the same as the log section above. 
It is always useful to have a section dedicated to miscellaneous information. 

In the case of feature requests, this could be used to point to relevant references, documentation, research papers, etc...
In the case of bugs, this section could help document information not captured above. 

Perhaps, we should rename it to miscellaneous information ?

File: .github/ISSUE_TEMPLATE.md (hunk 6 above)
Comment (richardpen) [inline .github/ISSUE_TEMPLATE.md]: Also we may want to provide an email here, so that if customer doesn't want publish all the logs publicly, they can send to this email.

File: .github/ISSUE_TEMPLATE.md (hunk 1 above)
Comment (richardpen) [inline .github/ISSUE_TEMPLATE.md]: I think these are already covered in the ecs-logs-collector, do we miss something in the ecs-logs-collector?

File: .github/ISSUE_TEMPLATE.md (hunk 6 above)
Comment (vsiddharth) [inline .github/ISSUE_TEMPLATE.md]: @richardpen Are you suggesting that we have a generic email address to obtain the logs from the customer ?

File: .github/ISSUE_TEMPLATE.md (hunk 1 above)
Comment (vsiddharth) [inline .github/ISSUE_TEMPLATE.md]: @richardpen The logs collector does most of this but we should still draw out the relevant sections here. 
This would also help customers who can't run the ecs-logs-collector.

File: .github/ISSUE_TEMPLATE.md (hunk 6 above)
Comment (richardpen) [inline .github/ISSUE_TEMPLATE.md]: yeah, that's what I can think of, unless you have better idea?

Comment (vsiddharth) [inline .github/ISSUE_TEMPLATE.md]: I'm in favor but we should be wary of getting spammed and getting hit with a lot of useless email. 
Perhaps we should think of a secure way to let customers share their logs with us besides email.

File: .github/ISSUE_TEMPLATE.md (hunk 1 above)
Comment (petderek) [inline .github/ISSUE_TEMPLATE.md]: Agreed, but we should definitely include examples of useful details

File: .github/ISSUE_TEMPLATE.md (hunk 6 above)
Comment (petderek) [inline .github/ISSUE_TEMPLATE.md]: Provide an address now; but we can be ready to select a better system if spam bots find the email.

File: .github/ISSUE_TEMPLATE.md (hunk 5 above)
Comment (samuelkarp) [inline .github/ISSUE_TEMPLATE.md]: I would prefer that we go without it.  If we decide we want this later, we can always add it.

File: .github/ISSUE_TEMPLATE.md (hunk 7 above)
Comment (samuelkarp) [inline .github/ISSUE_TEMPLATE.md]: Stack traces are in the logs.  So if you're attaching logs, they'll have the stack traces.

**Comment by vsiddharth** [comment]
I've updated the template with examples and reordered certain blocks. 
Please have another look. 

Thanks

File: .github/ISSUE_TEMPLATE.md [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+<!-- Please provide a brief outline of the issue -->
+
+
+### Description
+<!-- Provide detailed information about this issue -->
+
+
+
+<!--
+Please use these blocks to comment out appropriate sections
```
Comment (petderek) [inline .github/ISSUE_TEMPLATE.md]: I didn't understand this line at first. It may be clearer to put notes in sections we don't need.

&lt;!-- Not required for feature requests --&gt;

File: .github/ISSUE_TEMPLATE.md [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+especially for feature requests
+-->
+### Expected Behavior
+
+
+### Environment Details
```
Comment (petderek) [inline .github/ISSUE_TEMPLATE.md]: This is great!

File: .github/ISSUE_TEMPLATE.md [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+
+### Supporting Log Snippets
+<!--
+Please have a look at https://github.com/awslabs/ecs-logs-collector
+for the data we typically require for investigation
+-->
+
+
+### Additional Information
```
Comment (petderek) [inline .github/ISSUE_TEMPLATE.md]: I still find this to be implicit. At most, this could maybe be a bullet in the description section.

&lt;!-- Provide detailed information about this issue, *including any miscellaneous resources that brought you to this recommendation* --&gt;

The template is a guide, not a constraint. Folks can still add additional sections if they want to.

File: .github/ISSUE_TEMPLATE.md [hunk 12]
```diff
@@ -0,0 +1,46 @@
+<!--
+For high severity issues, please contact AWS Support directly.
```
Comment (nmeyerhans) [inline .github/ISSUE_TEMPLATE.md]: I might phrase this as "urgent operational issues" instead. I'd also like to see a link to AWS Support contact info. If somebody is in operational pain, they'll appreciate not having to dig for such details themselves.

File: .github/ISSUE_TEMPLATE.md [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+### Description
+<!-- Provide detailed information about this issue -->
+
+
+
+<!--
+Please use these blocks to comment out appropriate sections
+especially for feature requests
+-->
+### Expected Behavior
```
Comment (nmeyerhans) [inline .github/ISSUE_TEMPLATE.md]: If we're going to have an "expected behavior" section, we should also have "actual behavior" or "observed behavior" as well.

File: .github/ISSUE_TEMPLATE.md [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+### Environment Details
+<!--
+Examples:
+* docker info
+* curl http://localhost:51678/v1/metadata
+* df -kh
+-->
+
+
+### Supporting Log Snippets
```
Comment (nmeyerhans) [inline .github/ISSUE_TEMPLATE.md]: It may be worthwhile to include a reminder that GitHub issues are public and that the customer should take care to exclude potentially sensitive data from logs before posting. We'll request additional logs via less public channels if necessary.

**Comment by vsiddharth** [comment]
Revised the template yet again to incorporate feedback. 
Could you please have another look ?

Thanks

File: .github/ISSUE_TEMPLATE.md [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+### Observed Behavior
+
+
+### Environment Details
+<!--
+Examples:
+* docker info
+* curl http://localhost:51678/v1/metadata
+* df -kh
```
Comment (jhaynes) [inline .github/ISSUE_TEMPLATE.md]: Unless I'm mistaken, `-k` and `-h` are mutually exclusive. Whichever flag comes last wins (similarly with `--block-size` and the `-H` flag). If what you're wanting is human readable output with a block size of 1k (vs 1024), use `-H`.

File: .github/ISSUE_TEMPLATE.md [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+
+
+### Observed Behavior
+
+
+### Environment Details
+<!--
+Examples:
+* docker info
+* curl http://localhost:51678/v1/metadata
```
Comment (jhaynes) [inline .github/ISSUE_TEMPLATE.md]: I might add a `-v` to the curl command.

File: .github/ISSUE_TEMPLATE.md (hunk 14 above)
Comment (aaithal) [inline .github/ISSUE_TEMPLATE.md]: On that note, `please remove sensitive data` is a very generic message. As a customer, it's very hard to parse Agent and other logs to determine if they contain any sensitive data. Should we provide more details here? Or should the guidance be that logs should be collected, but we'll provide a mechanism to share them with ECS team after the issue has been created?

File: .github/ISSUE_TEMPLATE.md [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+* df -kh
+-->
+
+
+### Supporting Log Snippets
+<!--
+Please have a look at https://github.com/awslabs/ecs-logs-collector
+for the data we typically require for investigation
+
+GitHub issues are public, please remove sensitive data from logs before posting
```
Comment (aaithal) [inline .github/ISSUE_TEMPLATE.md]: As per @nmeyerhans note above, `please remove sensitive data` is a very generic message. As a customer, it's very hard to parse Agent and other logs to determine if they contain any sensitive data. Should we provide more details here? Or should the guidance be that logs should be collected, but we'll provide a mechanism to share them with ECS team after the issue has been created?

File: .github/ISSUE_TEMPLATE.md (hunk 15 above)
Comment (vsiddharth) [inline .github/ISSUE_TEMPLATE.md]: Yep, this is just extracted from the `ecs-logs-collector`. 
I think `df -h` would suffice here

File: .github/ISSUE_TEMPLATE.md (hunk 16 above)
Comment (vsiddharth) [inline .github/ISSUE_TEMPLATE.md]: Sure

File: .github/ISSUE_TEMPLATE.md (hunk 17 above)
Comment (vsiddharth) [inline .github/ISSUE_TEMPLATE.md]: What is sensitive varies a lot. I'm not sure if we need to list all the sensitive information present in our logs within a template. Wouldn't it be too much work for customers to manually filter out all the sensitive data contained in the logs ?? 
Also, we don't have a well defined mechanism or system for customers to share logs with us yet besides email.

Comment (aaithal) [inline .github/ISSUE_TEMPLATE.md]: > Wouldn't it be too much work for customers to manually filter out all the sensitive data contained in the logs ??

That's exactly my point. Which is why I suggested the alternative of telling customers to collect logs, but wait for our instructions to share the same until we have a better way of sharing logs.

Comment (nmeyerhans) [inline .github/ISSUE_TEMPLATE.md]: In general agent doesn't log anything truly sensitive (e.g. we'll never log a private crypto key). However, some customers have their own definition of what is sensitive or (in)appropriate for public posting. For example, container instance IDs aren't sensitive, but some people choose to redact them when posting publicly anyway. Same with container image names or task definition names. My recommendation is simply that we remind the customer that the issue is public and that they should take care to post only what they're comfortable posting. They can choose to redact information that they're not comfortable sharing, as long as the message meaning is not lost, or post nothing at all and wait for us to contact them out-of-band.

Comment (vsiddharth) [inline .github/ISSUE_TEMPLATE.md]: I agree here but not all log lines contain sensitive data and for some issues, customers could go ahead and post them using this template. If they are unsure, they should just hold off and request alternative means for sharing the logs. The line added here should serve as a warning. 

If customers are just going to describe the issue without any supporting logs; its going to get extremely difficult to make meaningful inferences.

Comment (aaithal) [inline .github/ISSUE_TEMPLATE.md]: > If they are unsure, they should just hold off and request alternative means for sharing the logs. 

It's worth adding a line, which states something to that effect in the template IMO. Leaves less room for ambiguity

Comment (jhaynes) [inline .github/ISSUE_TEMPLATE.md]: As we discussed, lets use wording similar to what @nmeyerhans suggested. e.g. "Remember, Github issues are public. If you are not comfortable posting your logs here, please let us know and we can provide an alternative method"

Comment (vsiddharth) [inline .github/ISSUE_TEMPLATE.md]: Extended the current warning for log section. 
Please let me know if you need any additional changes.
